src/librustdoc/html/sources.rs

```diff
@@ -54,12 +54,16 @@ impl DocFolder for SourceCollector<'_, '_> {
             self.scx.include_sources = match self.emit_source(&filename) {
                 Ok(()) => true,
                 Err(e) => {
-                    println!(
-                        "warning: source code was requested to be rendered, \
-                         but processing `{}` had an error: {}",
-                        filename, e
+                    let mut warning = self.scx.tcx.sess.struct_span_warn(
```

Comment:
So do you think it's okay to make this an error? The question is whether we think it's okay to fail the whole run if there's an error reading the source files.

Reply:
People can always use `#![doc(html_no_source)]` if they can't fix the error for some reason. We could even suggest that in a help message like:

```
help: if you are unable to resolve this error, you can skip including sources using the `#![doc(html_no_source)]` attribute at the crate root
```

Reply:
I think we should make this a hard error, yeah, that keeps it consistent with other errors:
```
$ rustdoc internal.rs 
error: couldn't generate documentation: Permission denied (os error 13)
  |
  = note: failed to create or modify "doc/src/internal"
```
I don't know if we need a suggestion - it seems weird that we'd be able to render the rest of the documentation but not the sources. I'd be very surprised to see this error happen in practice.

Reply:
Ok, here's a way to replicate the error:
```
$ mkdir -p doc/src/x
$ chmod a-w doc/src/x
$ rustdoc x.rs
warning: source code was requested to be rendered, but processing `x.rs` had an error: "doc/src/x/x.rs.html": Permission denied (os error 13)
         skipping rendering of source code
```
Could you add a test in `src/test/run-make` that this exits with an error?

Reply:
I was having trouble getting the test to work properly (cc #83435); it was passing even without the `chmod` calls. We talked about it and decided that we should just skip the test since this is a pretty uncommon error.